trieste/utils/pareto.py

```diff
@@ -101,10 +108,155 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
+            lower = tf.gather_nd(pseudo_front, tf.stack((lower_idx, arr), -1))
+            upper = tf.gather_nd(pseudo_front, tf.stack((upper_idx, arr), -1))
+
+            test_accepted = self._is_test_required((upper - jitter) < front)
+            lower_result, upper_result = tf.cond(
+                test_accepted,
+                lambda: self._accepted_test_body(lower_result, upper_result, lower_idx, upper_idx),
+                lambda: (lower_result, upper_result),
+            )
+
+            test_rejected = self._is_test_required((lower + jitter) < front)
+            divide_conquer_cells = tf.cond(
+                tf.logical_and(test_rejected, tf.logical_not(test_accepted)),
+                lambda: self._rejected_test_body(
+                    cell, lower, upper, divide_conquer_cells, total_size, threshold
+                ),
+                lambda: divide_conquer_cells,
+            )
+
+            return divide_conquer_cells, lower_result, upper_result
+
+        _, lower_result, upper_result = tf.while_loop(
+            lambda divide_conquer_cells, lower_result, upper_result: divide_conquer_cells.shape[0]
+            > 0,
+            while_body,
+            loop_vars=[divide_conquer_cells, lower_result, upper_result],
+            shape_invariants=[
+                tf.TensorShape([None, 2, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+            ],
+        )
+        return BoundedVolumes(lower_result, upper_result)
+
+    def _accepted_test_body(self, lower_result, upper_result, lower_idx, upper_idx):
+        lower_result = tf.concat([lower_result, lower_idx[None]], axis=0)
+        upper_result = tf.concat([upper_result, upper_idx[None]], axis=0)
+        return lower_result, upper_result
+
+    def _rejected_test_body(
+        self,
+        cell: TensorType,
+        lower: TensorType,
+        upper: TensorType,
+        divide_conquer_cells: TensorType,
+        total_size: TensorType,
+        threshold: TensorType,
+    ):
+
+        divide_conquer_cells_dist = cell[1] - cell[0]
+        hc_size = tf.math.reduce_prod(upper - lower, axis=0, keepdims=True)
+
+        not_unit_cell = tf.reduce_any(divide_conquer_cells_dist > 1)
+        vol_above_thresh = tf.reduce_all((hc_size[0] / total_size) > threshold)
+        divide_conquer_cells = tf.cond(
+            tf.logical_and(not_unit_cell, vol_above_thresh),
+            lambda: self._divide_body(divide_conquer_cells, divide_conquer_cells_dist, cell),
+            lambda: tf.identity(divide_conquer_cells),
+        )
+        return divide_conquer_cells
+
+    def _divide_body(
+        self,
+        divide_conquer_cells: TensorType,
+        divide_conquer_cells_dist: TensorType,
+        cell: TensorType,
+    ):
+        edge_size, idx = tf.reduce_max(divide_conquer_cells_dist), tf.argmax(
+            divide_conquer_cells_dist
+        )
```

Comment:
```suggestion
        edge_size = tf.reduce_max(divide_conquer_cells_dist)
        idx = tf.argmax(divide_conquer_cells_dist)
```
perhaps, since you're already going over the line?

Reply:
yeah thats better